File: sdk/java/src/main/java/com/devinsight/DevInsight.java
package com.devinsight;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;

public final class DevInsight {

    private static final String DEFAULT_BASE_URL = "https://devinsight-api.onrender.com";
    private static volatile boolean initialized = false;

    private DevInsight() {}

    /** Reads DEVINSIGHT_API_KEY from environment and registers the crash handler. */
    public static synchronized void init() {
        String apiKey = System.getenv("DEVINSIGHT_API_KEY");
        String project = envOr("DEVINSIGHT_PROJECT", "unknown");
        if (apiKey != null && !apiKey.isBlank()) {
            init(apiKey, project);
        }
    }

    /** Registers the crash handler with an explicit API key. */
    public static synchronized void init(String apiKey, String projectName) {
        if (initialized) return;
        if ("0".equals(System.getenv("DEVINSIGHT_ENABLED"))) return;
        if (apiKey == null || apiKey.isBlank()) return;

        initialized = true;

        final String key     = apiKey;
        final String project = projectName != null ? projectName : "unknown";
        final String base    = envOr("DEVINSIGHT_BASE_URL", DEFAULT_BASE_URL).replaceAll("/+$", "");

        Thread.setDefaultUncaughtExceptionHandler((thread, throwable) -> {
            send(key, base, project, throwable);
            System.exit(1);
        });
    }

    private static void send(String apiKey, String baseUrl, String projectName, Throwable t) {
        try {
            StringBuilder sb = new StringBuilder();
            for (StackTraceElement el : t.getStackTrace()) {
                sb.append("\tat ").append(el).append("\n");
            }
            String body = "{"
                + "\"tipo\":\"silent_backend_error\","
                + "\"mensagem\":" + jsonStr(t.getMessage()) + ","
                + "\"contexto\":{"
                +   "\"project_name\":"    + jsonStr(projectName) + ","
                +   "\"exception_type\":"  + jsonStr(t.getClass().getName()) + ","
                +   "\"stack\":"           + jsonStr(sb.toString())
                + "}}";

            HttpClient client = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(10))
                .build();

            HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + "/v1/diagnosticos"))
                .header("Content-Type", "application/json")
                .header("Authorization", "Bearer " + apiKey)
                .timeout(Duration.ofSeconds(10))
                .POST(HttpRequest.BodyPublishers.ofString(body))
                .build();

            client.send(req, HttpResponse.BodyHandlers.discarding());
        } catch (Exception e) {
            System.err.println("[DevInsight] Falha ao enviar diagnóstico: " + e.getMessage());
        }
    }

    private static String envOr(String key, String fallback) {
        String v = System.getenv(key);
        return (v != null && !v.isBlank()) ? v : fallback;
    }

    private static String jsonStr(String value) {
        if (value == null) return "null";
        return "\"" + value
            .replace("\\", "\\\\")
            .replace("\"", "\\\"")
            .replace("\n", "\\n")
            .replace("\r", "\\r")
            .replace("\t", "\\t")
            + "\"";
    }
}

File: sdk/go/devinsight.go
// Package devinsight captures panics and reports them to the DevInsight API.
//
// Usage:
//
//	func main() {
//	    devinsight.Wrap(run)
//	}
//
//	func run() {
//	    // your app code
//	}
package devinsight

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"runtime/debug"
	"strings"
	"time"
)

var (
	apiKey      = os.Getenv("DEVINSIGHT_API_KEY")
	baseURL     = strings.TrimRight(getEnv("DEVINSIGHT_BASE_URL", "https://devinsight-api.onrender.com"), "/")
	projectName = getEnv("DEVINSIGHT_PROJECT", "unknown")
	enabled     = os.Getenv("DEVINSIGHT_ENABLED") != "0"
)

func getEnv(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}

// Wrap runs fn and captures any panic, reporting it to DevInsight before re-panicking.
// This is the required one-liner for Go apps — place it as the first call in main().
func Wrap(fn func()) {
	defer func() {
		if r := recover(); r != nil {
			stack := string(debug.Stack())
			send(fmt.Sprintf("%v", r), "panic", stack)
			panic(r) // re-panic: preserves original crash behavior
		}
	}()
	fn()
}

func send(message, exceptionType, stack string) {
	if !enabled || apiKey == "" {
		return
	}

	payload := map[string]interface{}{
		"tipo":     "silent_backend_error",
		"mensagem": message,
		"contexto": map[string]string{
			"project_name":   projectName,
			"exception_type": exceptionType,
			"stack":          stack,
		},
	}

	body, err := json.Marshal(payload)
	if err != nil {
		fmt.Fprintf(os.Stderr, "[DevInsight] Falha ao serializar payload: %v\n", err)
		return
	}

	client := &http.Client{Timeout: 10 * time.Second}
	req, err := http.NewRequest(http.MethodPost, baseURL+"/v1/diagnosticos", bytes.NewReader(body))
	if err != nil {
		fmt.Fprintf(os.Stderr, "[DevInsight] Falha ao criar requisição: %v\n", err)
		return
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+apiKey)

	resp, err := client.Do(req)
	if err != nil {
		fmt.Fprintf(os.Stderr, "[DevInsight] Falha ao enviar diagnóstico: %v\n", err)
		return
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		fmt.Fprintf(os.Stderr, "[DevInsight] Servidor retornou erro: status %d\n", resp.StatusCode)
	}
}
